### ISSUE TITLE
Overlapping instances for Data.Aeson.TH.LookupField with Maybe

### ISSUE THREAD

**Issue by spl** [bos/aeson#468]
I'm currently upgrading a package from `aeson-0.8.1.1` to `aeson-0.11.2.1` (and GHC 7.10.2 to 8.0.1), and I've run into an issue with TH that works in the former but not in the latter.

Consider the following code (assuming the appropriate language extensions):

```haskell
module A where
type family F :: * -> *
data T = T { f1 :: F (), f2 :: Maybe (F ()) }

module B where
import A
import Data.Aeson
import Data.Aeson.TH
instance FromJSON (F ()) => FromJSON T where
  parseJSON = $(mkParseJSON defaultOptions ''T)
```

I get the following error when compiling:

```
        • Overlapping instances for Data.Aeson.TH.LookupField (F ())
            arising from a use of ‘Data.Aeson.TH.lookupField’
          Matching instances:
            instance [overlappable] FromJSON a => Data.Aeson.TH.LookupField a
              -- Defined in ‘Data.Aeson.TH’
            instance FromJSON a => Data.Aeson.TH.LookupField (Maybe a)
              -- Defined in ‘Data.Aeson.TH’
          (The choice depends on the instantiation of ‘’
           To pick the first instance above, use IncoherentInstances
           when compiling the other instance declarations)
        • In the second argument of ‘(<$>)’, namely
            ‘(Data.Aeson.TH.lookupField
                "A.T"
                "T"
                recObj_aauM
                (text-1.2.2.1:Data.Text.pack "f1"))’
          In the first argument of ‘(<*>)’, namely
            ‘(S.T
              <$>
                (Data.Aeson.TH.lookupField
                   "A.T"
                   "T"
                   recObj_aauM
                   (text-1.2.2.1:Data.Text.pack "f1")))’
          In the expression:
            ((S.T
              <$>
                (Data.Aeson.TH.lookupField
                   "A.T"
                   "T"
                   recObj_aauM
                   (text-1.2.2.1:Data.Text.pack "f1")))
             <*>
               (Data.Aeson.TH.lookupField
                  "A.T"
                  "T"
                  recObj_aauM
                  (text-1.2.2.1:Data.Text.pack "f2")))
```

The error appears to be specific to having at least two field names, one of them a `Maybe` and one of them not. It doesn't happen for the following definitions of `T`:

```haskell
data T = T (F ()) (Maybe (F ()))
data T = T { f1 :: F () }
data T = T { f1 :: F (), f2 :: F () }
data T = T { f1 :: Maybe (F ()) }
data T = T { f1 :: Maybe (F ()), f2 :: Maybe (F ()) }
data T = T { f1 :: F (), f2 :: Either () (F ()) }
```

The error appears to be specific to both type and data families. (It still occurs if you change `F` to a `data family`.) It does not happen for `F` defined as follows:

```haskell
data F a
newtype F a = F ()
```

**Comment by bergmark** [comment]
I'm guessing the instances used to be incoherent before 0.11 as all instances were marked as such. A sad consequence of the module level pragmas, and not something I want to re-introduce. But we didn't make this case break on purpose.

Is there a way to fix this without re-introducing IncoherentInstances? @RyanGlScott is this something you could take a look at? Thanks :)

**Comment by RyanGlScott** [comment]
I'm not knowledgeable enough on how one can remove `OverlappingInstances` in this sort of scenario. It's tricky because you need a form of _ad hoc_ polymorphism that does one thing for `Maybe a`, but something else for every other type `a` (without declaring a `LookupField` instance for every other type in existence, obviously). This might be possible with open type families (added in GHC 7.8), but since `aeson` supports back to GHC 7.4, we can't use that here.

FWIW, the GHC generics–based implementation also suffers from this issue, so that's not a workaround.

@phadej, perhaps you'd have some ideas as the author of [this gist](https://gist.github.com/phadej/cae76444548b9cffa10d9675e013847b)?

**Comment by spl** [comment]
@RyanGlScott:

> This might be possible with open type families (added in GHC 7.8), but since `aeson` supports back to GHC 7.4, we can't use that here.

I assume you meant _closed_ type families. That thought had occurred to me, too.

Assuming a solution does work only in newer versions of GHC, would it be worth considering having two different approaches distinguished with CPP? The benefit would be supporting more types in newer versions. The drawback would be a slightly higher maintenance burden perhaps? But at least the solution would not introduce differences in the runtime of programs using different versions of GHC.

**Comment by spl** [comment]
@bergmark: I've confirmed that `IncoherentInstances` fixes this in #470. All Travis-CI tests passed, so I assume that provides some indication that it's fixed for all supported GHC versions.

That PR includes a test that triggers the issue. It turns out that a type family is not required. Only a type parameter is needed:

```haskell
data T a = T { f1 :: a, f2 :: Maybe a }
```

There may be a way to do this without `IncoherentInstances`. I haven't looked into any, yet. But having #470 should make it easy for somebody to try.

If you're concerned that having `{-# LANGUAGE IncoherentInstances #-}` might “infect” other instances in `Data.Aeson.TH.hs` (which, I don't think, currently has any other instances), then one option would be to put just the `LookupField` class and instances in a separate module, say `Data.Aeson.TH.LookupField.hs`, to reduce the scope of the language extension.

**Comment by RyanGlScott** [comment]
@spl: Oops, I meant to type "closed" instead of "open". But in any case, I just tried out a closed-type-families–based solution, but it still suffers from the same issue. Unless someone exceedingly clever suggests otherwise, I can't see a solution to this problem that doesn't require `IncoherentInstances`.

**Comment by phadej** [comment]
@RyanGlScott I think that `Data.Type.Equality` trick might work, but it eventually relies on closed type families as well. So even possible: will mean narrowing support window.

**Comment by phadej** [comment]
can't `TH` code be done without `LookupField`, e.g. using `reify`? That won't fix `Generics` -implementation, but still better than nothing?

**Comment by RyanGlScott** [comment]
> I think that `Data.Type.Equality` trick might work, but it eventually relies on closed type families as well. So even possible: will mean narrowing support window.

I did try a `:~:`-based trick to check if `a` was equal to `Maybe b` (for some `Typeable b`), but then GHC complains that `b` is ambiguous... but perhaps I'm doing it wrong. :)

> can't `TH` code be done without `LookupField`, e.g. using `reify`?

Sort of. We know the reified `Type`s of each constructor field, but there are certain corner cases that would give this code trouble. For example, consider:

```haskell
type family Id a where
  Id a = a

data Foo a = Foo { f1 :: Id a, f2 :: Id (Maybe a) }
```

With the current TH implementation (using typeclasses), GHC can easily figure out that calling `lookupField` on `f2` should dispatch to the `LookupField (Maybe a)` instance. If we determined which variant of `lookupField` to call using just the reified `Type`, it'd mistakenly choose the catch-all case instead of the `Maybe a` case since the type is `Id (Maybe a)` (i.e., the outermost type constructor is not `Maybe`). That is, the TH code can only expand type _synonyms_, not type _families_.

That isn't to say that expanding type families using Template Haskell isn't impossible, but it adds a significant amount of complexity and doesn't always work. The `th-desugar` library's [`Language.Haskell.TH.Desugar.Expand`](http://hackage.haskell.org/package/th-desugar-1.6/docs/Language-Haskell-TH-Desugar-Expand.html) module (see its implementation [here](http://hackage.haskell.org/package/th-desugar-1.6/docs/src/Language.Haskell.TH.Desugar.Expand.html)) attempts to expand type families with the [`expand`](http://hackage.haskell.org/package/th-desugar-1.6/docs/Language-Haskell-TH-Desugar-Expand.html#v:expand) function, but with the caveat that it doesn't work for type families that distinguishes cases based on kind.

**Comment by spl** [comment]
The situation in which I need this is with an open type family.

**Comment by RyanGlScott** [comment]
@spl: Right, that's the point. The `reify` trick would allow `Data.Aeson.TH` to work with unresolved open type families like `F`, but it would _introduce_ a regression for resolvable, closed type families like `Id`, which is why I'm hesitant to do something like that.

**Comment by phadej** [comment]
I'm curious, @spl, the type families + `Maybe` stuff is really pushing what one can do with generics / th atm. Is it something you really need to rely on them, isn't writing instance by hand work for you?

I.e. I'm not sure how hard, or at all, we should try to fix this.

**Comment by spl** [comment]
@RyanGlScott:

> The `reify` trick would allow `Data.Aeson.TH` to work with unresolved open type families like `F`, but it would _introduce_ a regression for resolvable, closed type families like `Id`, which is why I'm hesitant to do something like that.

Ah, got it.

By the way, while we're talking about regressions, I don't see a test for a data type with a closed type family like `Id`. Should I add one?

@phadej:

> I'm curious, @spl, the type families + `Maybe` stuff is really pushing what one can do with generics / th atm. Is it something you really need to rely on them, isn't writing instance by hand work for you?

First, as I mentioned above (https://github.com/bos/aeson/issues/468#issuecomment-250404432), the problem here is not restricted to type families alone. A parameterized record type with a type parameter `a`, a field of type `a`, and a field of type `Maybe a` is enough to trigger it:

```haskell
data T a = T { f1 :: a, f2 :: Maybe a }
```

Second, the type family trick is absolutely something I'm relying on, and I've got way too many types using this to write instances for each by hand. Let me explain my situation. 😄

I've got 3 packages (in a simplified version of the story):

* `server`: compiled by GHC
* `client`: compiled by GHCJS
* `shared`: a common dependency for `server` and `client` that specifies a GADT request/response interface as well as JSON-serialized types used in that interface

I want to share database ID types between the `server` and `client`. That is, for a type `User`, I want a type `UserId` (which is not an `Int` but may wrap one) that can be passed via the request/response interface. On the `server`, I'm using `persistent`, which has its own way of specifying database ID types. So I want to tie that type, via JSON, to another type on the `client` without having to hand-code the conversion (which is a huge pain for a large interface that is constantly under development).

The solution I have at the moment:

* `shared`:

  ```haskell
  type family Id :: * -> *
  data User_
  newtype User { userId :: Id User_ }
  instance FromJSON (Id User_) => FromJSON User where
    parseJSON = $(mkParseJSON defaultOptions ''User)
  ```

* `client`:

  ```haskell
  newtype UserId = UserId Int
  deriveJSON defaultOptions ''UserId
  type instance Id User_ = UserId
  ```

* `server`:

  ```haskell
  type instance Id User_ = UserId
  -- ^ generated by persistent, includes FromJSON and ToJSON instances
  ```

Now, I didn't mention `Maybe` above, but I've got other types like this:

```haskell
data UserAction = UserAction
  { who    :: UserId
  , toWhom :: Maybe UserId
  , what   :: Action
  }
```

And the request/response GADT looks like this:

```haskell
data Req :: * -> * where
  DoAction :: UserAction -> Req ActionResult
```

In conclusion, using the (to steal @RyanGlScott's words) unresolved open type family `Id` is extremely helpful in specifying a type-safe interface in which I can rely on `aeson` to translate the representation and not have to convert between the ID types myself.

**Comment by RyanGlScott** [comment]
Alright, I have a third idea for a solution that would eliminate the use of `IncoherentInstances` (and some uses of `OverlappingInstances`). Ultimately, we're always calling `lookupField` on values of type `Value -> Parser a`. But these values are produced via `FromJSON` instances! This raises an interesting idea: what if we added another internal method to `FromJSON`?

```haskell
class FromJSON a where
  parseJSON :: Value -> Parser a
  ...
  lookupField :: (Value -> Parser a) -> String -> String -> Object -> Text -> Parser a
  lookupField = lookupFieldWith
```

We needn't export `lookupField` from, say, `Data.Aeson`, but it could be used internally. For most `FromJSON` instances, `lookupField` would do the "normal" thing (`lookupFieldWith`). But for the `FromJSON (Maybe a)` instance only, we define `lookupField pj _ _ = parseOptionalFieldWith`!

Similarly, we'd need to add a `liftLookupField :: (Value -> Parser (f a)) -> String -> String -> Object -> Text -> Parser (f a)` method to `FromJSON1` for when we're deriving `FromJSON1` values, and change the `FromJSON1 Maybe` instance accordingly. We'd also need:

```haskell
liftLookupField2 :: FromJSON2 f => (Value -> Parser (f a b)) -> String -> String -> Object -> Text -> Parser (f a b)
liftLookupField2 = lookupFieldWith
```

(This doesn't need to be a method of `FromJSON2` itself, since there's no `FromJSON2` instance for `Maybe`, so it will always do the right thing.)

Then in the TH code, we'd choose between `lookupField`, `liftLookupField`, and `liftLookupField2` accordingly depending on the type of a constructor's arguments. We can employ a similar trick for the generics code.

Is there any reason this couldn't work? I'm going to try mocking this up to see if it actually works in practice, and not just in the ideal world of my imagination :)

**Comment by spl** [comment]
@RyanGlScott: Sounds quite reasonable.

**Comment by RyanGlScott** [comment]
Argh, it turns out there _is_ a reason this can't work. This approach falls apart when you try to derive `FromJSON1` for something like:

```haskell
data T a = T { f1 :: Int, f2 :: a }
```

For `f1 :: Int`, we have that `Int` is a `FromJSON` instance, so can just invoke `FromJSON.lookupField` and we're good to go. But what about `f2 :: a`? Since we're deriving `FromJSON1`, we don't call `parseJSON` to parse a value of type `a`, but instead we invoke the argument to `liftParseJSON` of type `Value -> Parser a`. But this doesn't assume `a` is an instance of `FromJSON`! So we're back to square one again, since we can't distinguish between values of type `Maybe a` and other types using the `FromJSON` machinery. (For example, we'd need some way to behave differently when parsing values of type `T Bool` versus values of type `T (Maybe Bool)`!)

**Comment by phadej** [comment]
@RyanGlScott https://github.com/futurice/haskell-futurice-prelude/pull/79/files, it's not particularly elegant but it works without overlapping instances!
